packages/protocol/contracts/identity/FederatedAttestations.sol

```diff
@@ -0,0 +1,160 @@
+    _transferOwnership(msg.sender);
+    setRegistry(registryAddress);
+    // TODO ASv2 initialize any other variables here
+  }
+
+  /**
+   * @notice Returns the storage, major, minor, and patch version of the contract.
+   * @return The storage, major, minor, and patch version of the contract.
+   */
+  function getVersionNumber() external pure returns (uint256, uint256, uint256, uint256) {
+    return (1, 1, 0, 0);
+  }
+
+  function _isRevoked(address signer, uint256 time) internal returns (bool) {
+    if (revokedSigners[signer] > 0 && revokedSigners[signer] >= time) {
+      return true;
+    }
+    return false;
+  }
+
+  function lookupAttestations(string identifier, address[] trustedIssuers)
+    public
+    view
+    returns (Attestation[])
+  {
+    Attestation[] memory attestations = new Attestation[];
+    for (uint256 i = 0; i < trustedIssuers.length; i++) {
+      address trustedIssuer = trustedIssuers[i];
+      for (uint256 j = 0; j < identifierToAddresses[identifier][trustedIssuer].length; j++) {
+        Attestation memory attestation = identifierToAddresses[identifier][trustedIssuer][j];
+        if (!_isRevoked(attestation.signer, attestation.issuedOn)) {
+          attestations.push(attestation);
+        }
+      }
+    }
+    return attestations;
+  }
+
+  function lookupIdentifiersbyAddress(address account, address[] trustedIssuers)
+    public
+    view
+    returns (bytes32[])
+  {
+    bytes32[] memory identifiers = new bytes32[];
+    for (uint256 i = 0; i < trustedIssuers.length; i++) {
+      address trustedIssuer = trustedIssuers[i];
+      for (uint256 j = 0; j < addressToIdentifiers[account][trustedIssuer].length; j++) {
+        bytes32 memory identifier = addressToIdentifiers[account][trustedIssuer][j];
+        Attestation memory attestation = identifierToAddresses[identifier][trustedIssuer];
+        if (!_isRevoked(attestation.signer, attestation.issuedOn)) {
+          identifiers.push(identifier);
+        }
+      }
+    }
+    return identifiers;
+  }
+
+  function validateAttestation(
+    bytes32 identifier,
+    address issuer,
+    Attestation attestation,
+    uint8 v,
+    bytes32 r,
+    bytes32 s
+  ) public view returns (address) {}
+
+  function registerAttestation(bytes32 identifier, address issuer, Attestation attestation) public {
+    require(
+      msg.sender == attestation.account || msg.sender == issuer || msg.sender == attestation.signer
+    );
+    for (uint256 i = 0; i < identifierToAddresses[identifier][issuer].length; i++) {
+      require(identifierToAddresses[identifier][issuer][i] != attestation.account);
+    }
+    identifierToAddresses[identifier][issuer].push(attestation);
+    addressToIdentifiers[attestation.account][issuer] = identifier;
+  }
+
+  function deleteAttestation(bytes32 identifier, address issuer, address account) public {
+    require(msg.sender == attestation.account || msg.sender == issuer);
```

Comment:
instead of checking against the `issuer`, should we check that `msg.sender` is an authorized signer for `issuer`? Otherwise we require issuers to use their issuer key instead of delegating this to the signers, which could be insecure